### PR TITLE
KAFKA-6028: Improve the quota throttle communication (KIP-219)

### PR DESCRIPTION
**Pull request by jonlee2**
This PR is a continuation from https://github.com/apache/kafka/pull/4830. I create a new PR because the original PR got big and it looks like some reviewers are having problems with loading it. I also used a new branch. 

This implements KIP-219, where a broker return a response with throttle time on quota violation immediately after processing the corresponding request. After the response is sent out, the broker will keep the channel muted until the throttle time is over. Also, on receiving a response with throttle time, client will also block outgoing communication to the broker for the specified throttle time.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jonlee2** [comment]
@rajinisivaram @lindong28 FYI, this is a new PR that I'll use from now.

**Comment by jonlee2** [comment]
@cmccabe Hi Colin. Could you do a quick review on FetchSessionHandler.java changes I made in this PR?

With this KIP, brokers will send empty fetch responses with INVALID_SESSION_ID on fetch quota violation. I modified FetchSessionHandler.java so that the handler can distinguish these empty responses and keep the current session if any ongoing.

Thank you.
Jon

**Comment by jonlee2** [comment]
@cmccabe Actually, after further discussion with @lindong28, we decided to address the incremental fetch session id in a separate PR after submitting this one. The current behavior is that a full fetch is performed after throttling.

**Comment by lindong28** [comment]
@rajinisivaram Jon said that the failed system tests can pass now. Do we need to run the system test using that URL, or is it OK if the system test passes on one's desktop?

Also, one caveat of the current patch is that, if the FetchRequest is throttled, we will send a FetchResponse with INVALID_SESSION_ID, which causes client to reset its session and possibly send a full fetch request next time. It does not affect correctness and it affects performance when the client is throttled. So it maybe OK to address it in a separate patch, in which we can discuss with Colin etc. on the proper solution. Do you think it is OK?

**Comment by lindong28** [comment]
@jonlee2 Could you run all systems tests on your desktop?

**Comment by lindong28** [comment]
Hey @junrao, for large patch like this, would you like to review it, or is it OK to merge it after review by @rajinisivaram and I?

**Comment by lindong28** [comment]
Note that most of the comments for this patch can be found in https://github.com/apache/kafka/pull/4830.

**Comment by jonlee2** [comment]
@lindong28 I started system tests on my local box.

**Comment by rajinisivaram** [comment]
@jonlee2 @lindong28 Started system tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1756/

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Perhaps it would be better if `KafkaChannel.unmute` returned a boolean. Then you could do `if (channel.unmute())` instead of two calls. You could also change the method name to `maybeUnmute` to make it more obvious.

**Comment by jonlee2** [comment]
@lindong28 The local system test run finished and there were 10 failures. 8 of them were "Unable to parse aggregate performance statistics on node 1: xxxx", and this was because some occasional stats dumped by ProducerPerformance.Stats.record() are not parseable by the test framework. Not related my changes.

The other two were an SSH exception and ducktape service nodes not going away after 10 mins. I reran them and they were gone. 

@rajinisivaram I still can't access the jenkins link. Can you grant me an access (or whom should I ask)?

**Comment by lindong28** [comment]
@jonlee2 That link appears to be a private link hosted by Confluent so there is probably not guarantee that it will be provided to anyone outside Confluent. They nicely give me the read access long time ago. Thanks @rajinisivaram for running the tests. I will help fetch the test result for this patch.

**Comment by lindong28** [comment]
All tests in https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1756/ have passed.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java
Comment (jonlee2) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Done

**Comment by jonlee2** [comment]
@rajinisivaram @lindong28 Addressed the last code comment.

**Comment by rajinisivaram** [comment]
@jonlee2 @lindong28 Colin is on vacation, so I think it makes sense to work on fetch session for throttled requests in a follow-on KIP. We definitely want to get this in before feature freeze on 29th May, but we will have a bit more time to merge smaller follow-on PRs if required.

**Review by rajinisivaram — APPROVED** [review]
@jonlee2 @lindong28 Thanks for all your work on this PR. LGTM.

**Comment by jonlee2** [comment]
@rajinisivaram @lindong28 Thanks you very much for reviewing it. Really appreciate it.

**Comment by rajinisivaram** [comment]
@jonlee2 Have you run this PR with realistic workloads at LinkedIn, particularly the one that was discussed in the KIP motivation section? It will be good to know if this PR fixes that issue. Thanks.

**Comment by jonlee2** [comment]
@rajinisivaram No. We'll need to deploy this PR to test that, but it may take a while until the Kafka release used in LinkedIn catches up.

If you don't have any further questions, I'll ask @lindong28 to merge this PR. Please let me know. Thank you.

**Comment by jonlee2** [comment]
@rajinisivaram @lindong28 I filed two follow-up tickets:
- [KAFKA-6944](https://issues.apache.org/jira/browse/KAFKA-6946) - Add system tests testing the new throttling behavior using older clients/brokers  
- [KAFKA-6966](https://issues.apache.org/jira/browse/KAFKA-6946) - Keep the session id for incremental fetch when fetch responses are throttled

**Comment by rajinisivaram** [comment]
@jonlee2 If you have a test cluster against which you run load tests, you could perhaps deploy with this PR and test it out?

**Comment by ijuma** [comment]
Yes, it would be great to do some load testing with quotas enabled to verify that this new approach indeed fixes the problem mentioned in the KIP. And that it doesn't cause other problems.

**Comment by jonlee2** [comment]
@rajinisivaram @ijuma  I've been trying to set up some load tests using our dev test cluster. I've done some verification and so far the patch seems working as intended, but I will need more runs to know better. 

I'll be on vacation starting from this Saturday. I'll try my best to get some results before I leave.

**Comment by rajinisivaram** [comment]
@jonlee2 Thank you for running the tests. Any status you can provide before you go on vacation would be useful. We should try and merge this soon since this is the largest PR for 2.0.0.

**Comment by jonlee2** [comment]
I've run some load tests to verify the behaviors of this patch. My findings are below.

1. To simulate the scenario mentioned in the motivation section of KIP-219, I set up the following:
(Here, “old” means the latest trunk, and “new” means this patch)
- 1 MB/s produce bandwidth quota
- max.in.flight.requests.per.connection = 1
- 1 old broker and 5 old clients
- 1 new broker and 5 new clients

In this setup, the old clients talk to the old broker only, and the new clients talk to the new broker only.

Then, using kafka-producer-perf-test.sh, I created workloads so that each broker would throttle over 1 second most of the time. And I set "request.timeout.ms" for one of the clients in each group to 1 second and observed their behaviors.
 
The old client went into a pathological loop of “send request” -> “disconnect due to timeout” -> “reconnect” -> “send request” -> "disconnect".  The new client received a response right away on quota violation and didn’t communicate with the broker until the throttling time was over. No disconnection happened.

2. new client - old broker
In this setup, I verified double throttling from both client and broker didn't happen. On quota violation, the client received the response after the throttling was over from the broker and once it received the response, it was able to send the next request right away.

3. old client - new broker
In this case, the old client received a throttled response right away and could also send another request right after that. However, due to the broker side throttling, it received the response for that request after the broker was done throttling. 

I noticed that batch expiration happened in some cases, which in my understanding is expected because they were sitting in RecordAccumulator while throttling is going on. While reading the related code, it was noted that when determining whether or not to expire explicitly muted topic partitions, we need to take the throttling time into account as well (i.e., the throttling time shouldn't be counted toward expiration timeout). @lindong28 kindly provided a patch for this while I'm working on the tests. I'll push it shortly.

Please let me know if these tests should be sufficient. Thanks.

**Comment by ijuma** [comment]
Thanks @jonlee2. Do any of the scenarios verify that the backpressure via socket channel muting works as expected (i.e. if a client doesn't respect the throttling time)? I wasn't sure if 3 did that.

**Comment by jonlee2** [comment]
@ijuma, did you mean a scenario where an old client keeps sending requests and eventually gets a timeout error because the new broker keeps muting the channel? I didn't cover that in the tests above, but I just tried it by setting max.in.flight.requests.per.connection to 100 and request.timeout.ms to 3 secs in the old client-new broker setup. I observed a repeating pattern of clients sending multiple requests and then disconnecting from the broker due to timeout.

**Comment by jonlee2** [comment]
@lindong28 @rajinisivaram Pushed the changes for the muted batch expiration.

**Comment by jonlee2** [comment]
@rajinisivaram, should above tests be sufficient? Can we merge this PR? Thanks.

**Comment by ijuma** [comment]
> @ijuma, did you mean a scenario where an old client keeps sending requests and eventually gets a timeout error because the new broker keeps muting the channel? 

@jonlee2, yes, I think what you describe seems close enough. I wanted to make sure that if clients don't respect the throttle time, things work as expected. By the way, there's a merge conflict in the ConsumerCoordinator file.

**Comment by lindong28** [comment]
Since Jon is on vacation and both Jon and Kafka committers want to have KIP-219 in the Kafka 2.0 release, I rebased this patch onto latest trunk and opened PR in https://github.com/apache/kafka/pull/5094. If everthing looks, we can merge this PR using the same content in https://github.com/apache/kafka/pull/5094.

**Comment by lindong28** [comment]
@rajinisivaram @ijuma Thanks for the review!

It seems that only one stream related test failed in https://github.com/apache/kafka/pull/5094, which is likely not due to this patch. Do you think we can merge this patch for 2.0 release, or is more test/review required for this PR?

**Comment by lindong28** [comment]
Thanks for the patch @jonlee2! Thanks for the review @rajinisivaram @junrao @ijuma! This PR is merged to the trunk now with the latest change in https://github.com/apache/kafka/pull/5094.

**Comment by cmccabe** [comment]
Hi @jonlee2,

Sorry.  I was on vacation the last two weeks, so I didn't get a chance to review this earlier.  The change to FetchSession.scala seems reasonable overall.

However, it seems like it would be better to add a new method like getThrottledResponse to the FetchSessionContext.  This would allow us to increment the fetch session epoch when returning the empty throttling response.  That would prevent the fetch session dropping.

(As you noted, clearly, the fetch session dropping is bad, and will result in extra network traffic when the throttle engages -- the opposite of what we want here.)

I also think it's kind of unfortunate that we iterate over the full fetch response so many times.  We might do it once to count the size, another time to get the data, and a third time to downconvert.  In the future, we might want to consider being more efficient here-- although that's probably outside the scope of this PR.

**Comment by jonlee2** [comment]
Hi @cmccabe,

Thanks for the input. I already filed KAFKA-6946 to fix this issue and am currently working on it. My current plan is that when returning an empty response,
- for incremental fetch, increase the session epoch and return the actual session id in the response
- for all others (including full fetch), return INVALID_SESSION_ID

So I believe this is in line with what you suggested. Can you confirm that this would work?

I agree that multiple iterations over fetch response is inefficient. I did look into ways to make it more efficient while working on KIP-219, but the required changes seemed quite involved and as you pointed out it was out of the scope. I'll file a ticket for the improvement.

**Comment by jonlee2** [comment]
@cmccabe 

Actually, after reading the code further, I noticed that handleFetchRequest() in KafkaApis.scala:502 will create a new fetch context regardless of throttling, which in turn increases the session epoch for incremental fetch context (FetchSesson.scala:772). So it seems to me that all we need is to return the actual session id in the throttled response (and don't do anything extra regarding the epoch) if it is incremental fetch. Is this correct?

**Comment by cmccabe** [comment]
> Actually, after reading the code further, I noticed that handleFetchRequest() in KafkaApis.scala:502 will create a new fetch context regardless of throttling, which in turn increases the session epoch for incremental fetch context (FetchSesson.scala:772). So it seems to me that all we need is to return the actual session id in the throttled response (and don't do anything extra regarding the epoch) if it is incremental fetch. Is this correct?

Yes, I believe this is correct.

You want to do something very similar to what updateAndGenerateResponseData is doing.
Check the session epoch to make sure it wasn't incremented by some other request.  Then, if the epoch is correct, generate an empty response like you are currently doing.  From `IncrementalFetchContext#updateAndGenerateResponseData`:

```
  override def updateAndGenerateResponseData(updates: FetchSession.RESP_MAP): FetchResponse[Records] = {
    session.synchronized {
      // Check to make sure that the session epoch didn't change in between
      // creating this fetch context and generating this response.
      val expectedEpoch = JFetchMetadata.nextEpoch(reqMetadata.epoch())
      if (session.epoch != expectedEpoch) {
        info(s"Incremental fetch session ${session.id} expected epoch $expectedEpoch, but " +
          s"got ${session.epoch}.  Possible duplicate request.")
        new FetchResponse(Errors.INVALID_FETCH_SESSION_EPOCH, new FetchSession.RESP_MAP, 0, session.id)
     } else {
         ..generate and return the response...
     }
```

One approach would be to make this function take Option[FetchSession.RESP_MAP]  instead of just RESP_MAP.  Or you could create a new function just to generate the empty response.

**Pull request closed without merging**
